mos_tests/rabbitmq_oslo/test_restarts.py

```diff
@@ -0,0 +1,249 @@
+           "git clone {repo} && "
+           "cd {repo_path} && "
+           "pip install -r requirements.txt -r test-requirements.txt && "
+           "dpkg -i {pkg} || "
+           "apt-get -f install -y".format(**kwargs))
+    logger.debug('Install "oslo.messaging-check-tool" on controller')
+    return remote.check_call(cmd)
+
+
+def configure_oslomessagingchecktool(remote, ctrl_ips, nova_user, nova_pass,
+                                     conf_file_path):
+    """Write configuration file on controller
+    :param remote: SSH connection point to controller;
+    :param ctrl_ips: List of controllers IPs;
+    :param nova_user: Name of Rabbit admin;
+    :param nova_pass: Password of Rabbit admin;
+    :param conf_file_path: Path where config file will be written.
+    """
+    # Create config file for "oslo.messaging-check-tool"
+    configs = (
+        "[DEFAULT]\n"
+        "debug=true\n"
+        "[oslo_messaging_rabbit]\n"
+        "rabbit_hosts = {rabbit_hosts}\n"
+        "rabbit_userid = {nova_user}\n"
+        "rabbit_password = {nova_pass}\n".format(
+            rabbit_hosts=', '.join([(x + ':5673') for x in ctrl_ips]),
+            nova_user=nova_user,
+            nova_pass=nova_pass))
+    # Write config to file on controller
+    logger.debug('Write "oslo.messaging-check-tool" config file to controller')
+    with remote.open(conf_file_path, 'w') as f:
+        f.write(configs)
+
+
+def restart_rabbitmq_serv(env, remote=None, sleep=10):
+    """Restart rabbitmq-server service on one or all controllers.
+    After each restart, check that rabbit is up and running.
+    :param env: Environment
+    :param remote: SSH connection point to controller.
+        Leave empty if you want to restart service on all controllers.
+    :param sleep: Seconds to wait after service restart
+    """
+    # 'sleep' is to wait for service startup. I'll be also checked later
+    restart_cmd = 'service rabbitmq-server restart && sleep %s' % sleep
+    controllers = env.get_nodes_by_role('controller')
+    if remote is None:
+        # restart on all controllers
+        logger.debug('Restart RabbinMQ server on all controllers one-by-one')
+        for controller in controllers:
+            with controller.ssh() as remote:
+                remote.check_call(restart_cmd)
+                wait_for_rabbit_running_nodes(remote, len(controllers))
+    else:
+        # restart on one controller
+        logger.debug('Restart RabbinMQ server on one controller')
+        remote.check_call(restart_cmd)
+        wait_for_rabbit_running_nodes(remote, len(controllers))
+
+
+def num_of_rabbit_running_nodes(remote, timeout_min=5):
+    """Get number of 'running_nodes' from 'rabbitmqctl cluster_status'
+    :param remote: SSH connection point to controller.
+    :param timeout_min: Timeout in minutes to wait for successful cmd execution
+    """
+    timeout = time.time() + 60 * timeout_min
+    cmd = 'rabbitmqctl cluster_status'
```

Comment:
After investigation I can say that API is a good thing, but not in case of service restart.
It's behavior is not stable in compare with CLI commands.
CLI can produce output if primary controller out, but API can't:
```
root@node-2:~# rabbitmqctl cluster_status
Cluster status of node 'rabbit@messaging-node-2' ...
[{nodes,[{disc,['rabbit@messaging-node-2']}]},{alarms,[]}]

root@node-2:~# curl -u nova:Ajl9OxOMW2mgB7fZRZbH6aPu http://localhost:15672/api/overview
curl: (7) Failed to connect to localhost port 15672: Connection refused
```

About `rabbitmqctl cluster_status | grep running_nodes`:
it is also not a good solution, because there can be several lines. And the number of lines depends according to the number of nodes